src/cyw43_bthci_uart.c

```diff
@@ -0,0 +1,297 @@
+
+static inline void cyw43_bluetooth_put_le16(uint8_t *buf, uint16_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+}
+
+static inline void cyw43_bluetooth_put_le32(uint8_t *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+    buf[3] = val >> 24;
+}
+
+static void cyw43_bluetooth_wait_cts_low(uint32_t timeout_us) {
+    uint32_t start_us = cyw43_hal_ticks_us();
+    while (cyw43_hal_pin_read(CYW43_PIN_BT_CTS) != 0) {
+        if (cyw43_hal_ticks_us() - start_us > timeout_us) {
+            // Timeout.
+            break;
+        }
+        cyw43_delay_ms(1);
+    }
+}
+
+static int cyw43_bluetooth_hci_cmd_raw(size_t len, uint8_t *buf) {
+    cyw43_bluetooth_wait_cts_low(10000);
+
+    // Send the HCI command to the BT controller.
+    cyw43_hal_uart_write((void *)buf, len);
+
+    // Read in the response header from the BT controller.
+    for (size_t i = 0; i < 6; ++i) {
+        int c = cyw43_hal_uart_readchar();
+        if (c < 0) {
+            return -CYW43_ETIMEDOUT;
+        }
+        buf[i] = c;
+    }
+
+    // Expect a command complete event (event 0x0e).
+    if (buf[0] != 0x04 || buf[1] != 0x0e) {
+        CYW43_WARN("cyw43_bluetooth_hci_cmd_raw: unknown response: %02x %02x %02x %02x\n", buf[0], buf[1], buf[2], buf[3]);
+        return -CYW43_EIO;
+    }
+
+    // Read in the response payload from the BT controller.
+    size_t sz = buf[2] - 3;
+    for (size_t i = 0; i < sz; ++i) {
+        int c = cyw43_hal_uart_readchar();
+        if (c < 0) {
+            return -CYW43_ETIMEDOUT;
+        }
+        buf[i] = c;
+    }
+
+    return 0;
+}
+
+static int cyw43_bluetooth_hci_cmd(int ogf, int ocf, size_t param_len, const uint8_t *param_buf) {
+    uint8_t *buf = cyw43_hal_uart_cmd_buf;
+    buf[0] = 0x01;
+    buf[1] = ocf;
+    buf[2] = ogf << 2 | ocf >> 8;
+    buf[3] = param_len;
+    if (param_len) {
+        memcpy(buf + 4, param_buf, param_len);
+    }
+    return cyw43_bluetooth_hci_cmd_raw(4 + param_len, buf);
+}
+
+#if defined(CYW43_BT_UART_BAUDRATE_DOWNLOAD_FIRMWARE) || defined(CYW43_BT_UART_BAUDRATE_ACTIVE_USE)
+static int cyw43_bluetooth_hci_set_baudrate(uint32_t baudrate) {
+    uint8_t buf[6];
+    cyw43_bluetooth_put_le16(buf, 0);
+    cyw43_bluetooth_put_le32(buf + 2, baudrate);
+    return cyw43_bluetooth_hci_cmd(0x3f, 0x18, 6, buf);
+}
+#endif
+
+static int cyw43_bluetooth_download_firmware(const uint8_t *firmware, size_t firmware_len) {
+    cyw43_bluetooth_hci_cmd(0x3f, 0x2e, 0, NULL);
+
+    const uint8_t *firmware_max = firmware + firmware_len;
+    bool last_packet = false;
+    while (!last_packet && firmware < firmware_max) {
+        // Validate firmware packet.
+        if (firmware[1] != 0xfc) {
+            CYW43_WARN("cyw43_bluetooth_download_firmware: unexpected packet %02x\n", firmware[1]);
+            return -CYW43_EIO;
+        }
+
+        // Construct HCI header.
+        uint8_t *buf = cyw43_hal_uart_cmd_buf;
```

Comment:
This buffer can be replaced with a stack buffer `uint8_t cmd_buf[256+4]`. One less config option to worry about.

Reply:
OK, you're right, this buffer could be fully stack allocated.

I've reworked this now and made it so that you can provide an optional static buffer.  If that's not defined then this driver automatically stack allocates it.

Reply:
Note this buffer is also used in `extmod/nimble/hal/hal_uart.c` in a very similar way. Replacing that one too with a stack buffer could save data memory that was unrecoverable before.
https://github.com/micropython/micropython/blob/master/extmod/nimble/hal/hal_uart.c#L73

Reply:
Yes, we can fix that eventually.